Memory refactored

damiankowal opened the pull request:
I wanted to try using Backbone models and views, and was learning as I went along, so the result may be a little messy at this stage.

The most challenging part of was figuring out how to get different objects listening to different events. One difficult problem was this: when 2 tiles are matched successfully, they need to remain visible for a short period before being hidden again. If the game is listening for clicks at that time how can it be designed so that events are handled in the right order?

damiankowal committed:
Diagram of memory game play added

Before I start coding anything, I want to try to understand how the game play works, the logic and events involved. Making this diagram to visualize it is helpful to me.

damiankowal committed:
Update diagram to include the idea of a turn, and define the point at which a game starts

Turns may be useful in the future for keeping score: the score should probably reflect not just the time taken to complete the game, but also how many times tiles were flipped. In other words, does completing the game in fewer turns indicate greater skill? The start of the game is also moved so that it is registered after the first tile is flipped, not when the board first appears.

damiankowal committed:
Adding dependencies and organizing scripts in a separate directory

damiankowal committed:
Trying to set up a backbone app by following an example from http://addyosmani.github.io/backbone-fundamentals/

damiankowal committed:
Render a very basic set of tiles using backbone models and views, relying on the examples at http://addyosmani.github.io/backbone-fundamentals/

damiankowal committed:
Add tiles to initial values to make pairs

damiankowal committed:
Add CSS to give tile elements the appearance of squares on a grid

damiankowal committed:
Refine media queries to try to maintain square tile shape

damiankowal committed:
Use underscore shuffle method to randomize the order of tiles on the board

damiankowal committed:
Add handler function to hide and show tiles on click, and update CSS

damiankowal committed:
Create game view to handle game events

damiankowal committed:
A complete mess at this point, trying to figure out what's going on with event handling

damiankowal committed:
Tile click event successfully passed using a game controller (backbone event) object

The example from this article was used as a starting point: http://lostechies.com/derickbailey/2011/07/19/references-routing-and-the-event-aggregator-coordinating-views-in-backbone-js/

damiankowal committed:
Remove game view, it added unnecessary complexity

damiankowal committed:
Successfully counting flipped tiles in the tiles collection

damiankowal committed:
Define a function to compare tiles and invoke it when 2 tiles are flipped

damiankowal committed:
Add a conditional statement to the tile selection handler so that it stops if a flipped tile is clicked

damiankowal committed:
Successfully compare tile values to determine whether or not they match

damiankowal committed:
Successfully mark resolved tiles

The logic seems messy, though. Resolved tiles are flipped, because their value is showing. It might be cleaner to not mark resolved tiles as flipped.

damiankowal committed:
Fix incorrect ID in CSS

damiankowal committed:
Removing unused file

damiankowal committed:
Add functions to test whether or not the game is over when each turn is complete

damiankowal committed:
Removing unused lines

damiankowal committed:
Merge branch 'backbone-implementation' into gh-pages

Refactored to use Backbone models and views.

damiankowal committed:
Update gruntfile and package.json

afeld commented on index.html:
```diff
@@ -1,9 +1,22 @@
+</head>
+<body>
+    <div id="board"></div>
+    <script id="tileTemplate" type="text/template">
```
:sparkles:

afeld commented on js/app.js:
```diff
@@ -0,0 +1,18 @@
+            { value: 'F' }, { value: 'F' },
+            { value: 'G' }, { value: 'G' },
+            { value: 'H' }, { value: 'H' }
+        ]
```
Minor: Rather than needing this much copy-paste, you might consider having a single array of values (`['A', 'B', ...]`) that gets rebuilt into the structure you need.

afeld commented on js/collections/tiles.js:
```diff
@@ -0,0 +1,13 @@
+app.Tiles = Backbone.Collection.extend({
+
+    initialize: function( options ) {
+        this.model = app.Tile;
```
Out of curiosity, why set this in the constructor, rather than in the properties object?

http://backbonejs.org/#Collection-model

damiankowal replied:
I didn't realize that I could set it as a property (being new to Backbone, I thought it needed to be in the constructor for some reason). I tried updating it this way:
https://github.com/damiankowal/memory/commit/448ea0cdd796b61c5c9dee2957aa9c32c2d62d43

afeld commented on js/collections/tiles.js:
```diff
@@ -0,0 +1,13 @@
+    flippedCount: function() {
+        var flippedTiles = this.where({ flipped: true });
+        return flippedTiles.length;
+    }
```
:+1:

afeld commented on js/app.js:
```diff
@@ -0,0 +1,18 @@
+var app = app || {};
```
Smart!

afeld commented on js/views/board.js:
```diff
@@ -0,0 +1,86 @@
+    initialize: function( options ) {
+        this.collection = new app.Tiles( options.initialTiles );
+        _.bindAll( this, 'tileSelection' );
+        app.gameController.bind( 'tileSelection', this.tileSelection );
```
Since this event is the same as a `change:flipped` event getting fired on the `Tile` model, I'd probably favor just using that so that there's one fewer moving piece.  You can subscribe to that event on the collection:

http://backbonejs.org/#Collection

afeld commented on js/views/board.js:
```diff
@@ -0,0 +1,86 @@
+        var flippedTiles;
+        if ( tile.get( 'flipped' ) || tile.get( 'resolved' ) ) {
+            console.log( 'a flipped or resolved tile should be ignored' );
+            return;
```
Try to avoid mid-method `return` statements.  Makes it a bit harder to reason about the flow at a glance.

/cc @advanced-js/students

damiankowal replied:
I rewrote the conditional statement to eliminate the return statement:
https://github.com/damiankowal/memory/commit/0cbf65f1cab6b78ab69ec3ebd58a93393fee393c

afeld commented on js/views/board.js:
```diff
@@ -0,0 +1,86 @@
+            resultHandler = function( tile ) {
+                tile.toggleFlip();
+            };
+        }
```
Interesting approach here with creating different callbacks for the different scenarios... I feel like there's a more idiomatic way, but I don't necessarily dislike it.  Hmm.  Will think on it.

afeld commented on js/views/board.js:
```diff
@@ -0,0 +1,86 @@
+            resolved: false
+        });
+        console.log ( unresolvedTiles.length + ' to go' );
+        if ( unresolvedTiles.length === 0 ) {
```
I'd probably split this logic into an even more specific method like `areAllTilesMatched()` or something.

afeld commented on js/views/board.js:
```diff
@@ -0,0 +1,86 @@
+    },
+
+    endGame: function() {
+        console.log( 'you win!' );
```
Minor: the user probably won't have their console open, so you should output this another way.

afeld commented:
Great work :metal:

afeld commented:
Sorry, forgot to address your comments above:

> The most challenging part of was figuring out how to get different objects listening to different events.

Yep :wink: 

> If the game is listening for clicks at that time how can it be designed so that events are handled in the right order?

Often you'll need to disable/ignore certain events if the page is in a certain state, e.g. if you're waiting for a transition and need to put up a loading spinner, and want to stop the user from messing with it in the meantime.  A very real problem!

damiankowal committed:
Set model for tiles collection in the properties object, rather than in the constructor

damiankowal committed:
Rewrite conditional statement to avoid mid-method return statement (and also catch an incorrect variable declaration)

damiankowal committed:
Define a function which builds a set of tiles from an inital array of values

damiankowal committed:
Split checkGameStatus function into several more specific methods

damiankowal committed:
Add missing semicolon

damiankowal committed:
Rewrite handleTurn function, defining several inner functions which allow the logic of handleTurn to be written more clearly

damiankowal committed:
Deleting duplicate function definitions; the functions were moved under handleTurn so that they could be written without 'this' infront of them

damiankowal committed:
Make some tile- and tile-collection-related functions methods of those models rather than defining them in the board view, making it possible to define the tileSelection function more succinctly

damiankowal committed:
Break the handleTurn function down further by defining clearly named handlers for successful and unsuccessful matches

This allows what happens when a set of tiles is matched or not matched to kept within the handler function - I think this is cleaner and easier to read.

damiankowal committed:
Define flip( tile ) as a method since it is being used in multiple instances

damiankowal committed:
Trigger a custom event on the tile model when a tile is clicked, and have the BoardView listen for that event

This allows for the removal of the 'gameController', simplifying things. Also, the logic to determine whether or not a tile should be listened to is moved from the board view to the tile view

damiankowal committed:
Rename some functions and variables more clearly, clean up console log messages

damiankowal committed:
Create message model and views to display a message, and use the new view to show a message in the browser when the game is complete (rather than logging a message in the console).

damiankowal committed:
Create game model and set the display view to listen for changes on that model to render welcome and win messages

damiankowal committed:
Expression used for registering the start of a game was throwing an error in jshint; switch to conditional statement instead

damiankowal committed:
Change game view and message view listeners to only listen once for events to start and stop the game, hopefully makeing the functions more efficient; this way, the board view does not need to keep checking whether or not the game has started on each tile selection

damiankowal committed:
Remove message model and message view, the message display is fairly straightforward and does not require its own model

damiankowal committed:
Update CSS and HTML (scripts loaded) to reflect changes in message view and removal of message-related files

damiankowal committed:
Rename model and view instances to make it easier to differentiate them.

damiankowal committed:
Add reset button view and HTML element, and add methods to other views and models to enable restarting the game

damiankowal committed:
Trigger a restart event on the game model so the different views can listen to it, allowing the message view to display an appropriate restart message; change naming for winning the game from 'ended' to 'completed', since a restart is also ending the game but not completing it

damiankowal committed:
Define separate getShuffledItems method in board view, as it is a general piece of functionality which can be separated from the render method; clean up stray log messages

damiankowal committed:
Update events and listenters so that messages respond properly to restarting of the game

damiankowal committed:
Removing unnecessary function

damiankowal commented:
I've updated the memory game to address some of the issues raised after the initial pull request:
- A message view has been added to show feedback in the browser window, rather than in the JavaScript console, while the game is being played.
- The game can be restarted with a button, both during the game and when the game is completed.
- View methods have been rewritten, defining smaller, more specific functions to avoid the more complicated, nested functions in the initial version.

afeld commented:
Thanks for all the hard work! :clap:
